tlssock.c

```diff
@@ -214,6 +214,35 @@ send(int sockfd, const void *buf, size_t len, int flags)
   return tls_write(tls, buf, len);
 }
 
+int
+accept(int sockfd, struct sockaddr *addr, socklen_t *addrlen)
+{
+  return accept4(sockfd, addr, addrlen, 0);
+}
+
+int
+accept4(int sockfd, struct sockaddr *addr,
+        socklen_t *addrlen, int flags)
+{
+  int clientfd;
+
+  tls_auto_t *tls = NULL;
+
+  tls = idx_get(sockfd);
+
+  clientfd = NEXT(accept4)(sockfd, addr, addrlen, flags);
+  if(clientfd < 0 || !tls)
+    return clientfd;
+
+  tls = tls_new(clientfd, false);
+  if (!tls || !idx_set(clientfd, tls, NULL)) {
+    close(clientfd);
+    return -1;
+  }
+
+  return clientfd;
+}
```

Comment:
1. A reference to the `tls_t` from `idx_get()` is leaked on line 237.
2. Fetching the listening socket's tls_t before `accept4()` could potentially cause `errno` deviations from the system behavior. Therefore, `accept4()` should be called first.

Here's how I would implement this:

```c
int
accept4(int sockfd, struct sockaddr *addr, socklen_t *addrlen, int flags)
{
  int fd;

  fd = NEXT(accept4)(sockfd, addr, addrlen, flags);

  if (fd >= 0) {
    tls_auto_t *lis = NULL;

    lis = idx_get(sockfd);
    if (lis) {
      tls_auto_t *con = NULL;

      // TODO: implement tls_dup() to copy options from the listening tls_t.
      // NOTE: this requires that we add *_dup() functions to GnuTLS for all creds.
      con = tls_new(fd, tls_is_client(lis));
      if (!con || !idx_set(fd, con, NULL)) {
        close(fd);
        return -1;
      }
    }
  }

  return fd;
}
```